pex/pex_boot.py

```diff
@@ -0,0 +1,219 @@
+            pex=pex, installed_location=installed_location
+        )
+    )
+    __re_exec__(sys.executable, python_args, installed_location)
+
+
+def __maybe_run_venv__(
+    pex,  # type: str
+    pex_root,  # type: str
+    pex_hash,  # type: str
+    has_interpreter_constraints,  # type: bool
+    hermetic_venv_scripts,  # type: bool
+    pex_path,  # type: Tuple[str, ...]
+    python_args,  # type: List[str]
+):
+    # type: (...) -> Optional[str]
+
+    from pex.common import is_exe
+    from pex.tracer import TRACER
+    from pex.variables import venv_dir
+
+    venv_root_dir = venv_dir(
+        pex_file=pex,
+        pex_root=pex_root,
+        pex_hash=pex_hash,
+        has_interpreter_constraints=has_interpreter_constraints,
+        pex_path=pex_path,
+    )
+    venv_pex = os.path.join(venv_root_dir, "pex")
+    if not __SHOULD_EXECUTE__ or not is_exe(venv_pex):
+        # Code in bootstrap_pex will (re)create the venv after selecting the correct
+        # interpreter.
+        return venv_root_dir
+
+    TRACER.log("Executing venv PEX for {pex} at {venv_pex}".format(pex=pex, venv_pex=venv_pex))
+    venv_python = os.path.join(venv_root_dir, "bin", "python")
+    if hermetic_venv_scripts:
+        __re_exec__(venv_python, python_args, "-sE", venv_pex)
+    else:
+        __re_exec__(venv_python, python_args, venv_pex)
+
+
+def boot(
+    bootstrap_dir,  # type: str
+    pex_root,  # type: str
+    pex_hash,  # type: str
+    has_interpreter_constraints,  # type: bool
+    hermetic_venv_scripts,  # type: bool
+    pex_path,  # type: Tuple[str, ...]
+    is_venv,  # type: bool
+    inject_python_args,  # type: Tuple[str, ...]
+):
+    # type: (...) -> int
+
+    entry_point = None  # type: Optional[str]
+    __file__ = globals().get("__file__")
+    __loader__ = globals().get("__loader__")
+    if __file__ is not None and os.path.exists(__file__):
+        entry_point = __entry_point_from_filename__(__file__)
+    elif __loader__ is not None:
+        if hasattr(__loader__, "archive"):
+            entry_point = __loader__.archive
+        elif hasattr(__loader__, "get_filename"):
+            # The source of the loader interface has changed over the course of Python history
+            # from `pkgutil.ImpLoader` to `importlib.abc.Loader`, but the existence and
+            # semantics of `get_filename` has remained constant; so we just check for the
+            # method.
+            entry_point = __entry_point_from_filename__(__loader__.get_filename())
+
+    if entry_point is None:
+        sys.stderr.write("Could not launch python executable!\\n")
+        return 2
+
+    python_args = list(inject_python_args)  # type: List[str]
+    orig_args = orig_argv()
+    if orig_args:
+        for index, arg in enumerate(orig_args[1:], start=1):
+            if os.path.exists(arg) and os.path.samefile(entry_point, arg):
+                python_args.extend(orig_args[1:index])
+                break
```

Comment:
I added some manual timing around this during development - I was a bit worried about the ctypes impl since there is no opt-out even if you don't use Python args and this is in the hot path. The timings were consistently sub-ms for both sys.orig_argv and the ctypes implementation. Almost all the ~O(100us) time for the ctypes version was spent in the imports above where the function is defined. The call here was much faster.